report/3.Theory/TypeRules.tex

```diff
@@ -1,13 +1,373 @@
+% !TeX root = ../main.tex
 %Content includes all type rules in HCL
 %Should be a relatively short section, since much of the content is self explanatory.
 \section{Type Rules}
-%introduction
+HCL employs types.
+This is done for three reasons:
+\begin{itemize}
+	\item To avoid certain runtime errors, which would not be caught without types.
+	\item Easier allocation of memory, as each type has a pre-defined size.
+	\item As described in the problem analysis, types make it easier for novices to understand constructions such as function definitions.
+\end{itemize}
+The code snippet shown in listing \ref{lis:TypeEx} is a syntactically correct assignment in HCL.
+If type rules were not enforced, this program would fail (or run with unexpected results), as the second argument for the "$+$" function would evaluate to a value of type bool, and no definition for this function allows bool-type arguments.
+
+\begin{lstlisting}[language=HCL,label=lis:TypeEx,caption=An HCL expression,firstnumber=1]
+x = 10 + (2 equals 3)
+\end{lstlisting}
+
+To avoid this, the HCL compiler includes a type-checker which will be described in a later chapter.
+
+The HCL type system consists of three components:
+\begin{itemize}
+	\item The set of all types in HCL.
+	\item Type rules for each syntactic category.
+	\item A definition for each type rule.
+\end{itemize}
+
+Essentially, the type rules specify when certain expressions have to evaluate to specific types.
+The rules have been split up in two sections:
+\begin{itemize}
+	\item Rules for expressions
+	\item Rules for statements
+\end{itemize}
+The former specifies which type an expression evaluates to, depending on the input.
+The latter specifies what types the components of a statement must be to be type valid.
+
+\textbf{Notation}\\
+The notation for type rules is very similar to the notation of semantics with a few differences.
+Below is an example:\\
+\begin{center}
+	\begin{math}
+	\cfrac
+	{E \vdash e_1 : num \quad e_2 : num}
+	{E \vdash e_1\ + e_2 : num}
+	\end{math}
+\end{center}
+This is the type rule for addition of two numbers.
+It means that given the current environment, the resulting value will be of type $num$ if the two input expressions are both of type $num$.
+
+As the type rules do not change the state of a program, and therefore also do not change the variable- and function environments, the environment can be simplified to one symbol, $E$.
+The colon is used to show that an expression is of a certain type.
 
 \subsection{Type Rules for Expressions}
-%quick introduction.
+As mentioned earlier, an expression evaluates to a type, if certain premises are upheld (where such apply).
```

Comment:
omvendt ordstilling. 
"if certain premises are upheld, an expression evaluates to a type."
hele "where such apply" virker ligegyldigt her. not sure

Reply:
Jeg er virkelig ikke glad for at starte en subsection med "if".

Reply:
Jeg har skrevet "if and only if" i stedet for "if"